python/neuroglancer/pipeline/storage.py

```diff
@@ -38,59 +44,76 @@ def __init__(self, layer_path='', n_threads=20):
         self._n_threads = n_threads
 
         if self._path.protocol == 'file':
-            self._interface = FileInterface
+            self._interface_cls = FileInterface
         elif self._path.protocol == 'gs':
-            self._interface = GoogleCloudStorageInterface
+            self._interface_cls = GoogleCloudStorageInterface
         elif self._path.protocol == 's3':
-            self._interface = S3Interface
+            self._interface_cls = S3Interface
+
+        self._interface = self._interface_cls(self._path)
+
+        self._queue = Queue.Queue(maxsize=0) # infinite size
+        self._threads = ()
+        self._terminate = threading.Event()
 
-        if self._n_threads:
-            self._create_processing_queue()
+        self.start_threads(n_threads)
+
+    @property
+    def layer_path(self):
+        return self._layer_path
 
     def get_path_to_file(self, file_path):
         return os.path.join(self._layer_path, file_path)
 
-    def _create_processing_queue(self):
-        self._queue = Queue(maxsize=self._n_threads*4)
-        for _ in xrange(self._n_threads):
-            worker = Thread(target=self._process_task)
-            worker.setDaemon(True)
+    def start_threads(self, n_threads):
+        if n_threads == len(self._threads):
+            return self
+
+        self.kill_threads()
+        self._terminate = threading.Event()
+
+        threads = []
+
+        for _ in xrange(n_threads):
+            worker = threading.Thread(
+                target=self._consume_queue, 
+                args=(self._terminate,)
+            )
+            worker.daemon = True
             worker.start()
+            threads.append(worker)
 
-    def _kill_threads(self):
-        self.wait_until_queue_empty()
-        for _ in xrange(self._n_threads):
-            self._queue.put( ('TERMINATE', None, None ) )
+        self._threads = tuple(threads)
+        return self
 
-    def _process_task(self):
-        """
-        Connections to s3 or gcs are likely not thread-safe,
-        to account for that every worker create it's own 
-        connection.
-        """
-        interface = self._interface(self._path)
-        while True:
-            # task[0] referes to an string with the method name.
-            # task[1] is a callback
-            # task[2:] are the arguments to the method.
-            task = self._queue.get()
-            fn_name, cb, args = task[0], task[1], task[2:]
-            
-            if fn_name == 'TERMINATE':
-                self._queue.task_done()
-                return
+    def are_threads_alive(self):
+        return any(map(lambda t: t.isAlive(), self._threads))
+
+    def kill_threads(self):
```

Comment:
Unfortunately, there's a reason for this. In other code we often create lots of Storage instances and their threads hang around. They don't get garbage collected often enough 
so you really need to destroy them, and `del storage` is unreliable. Thus you have two options depending on your programming style:

1) Call storage.kill_threads() when you're done.
2) Use the `with` statement